src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java

```diff
@@ -651,10 +676,14 @@ private void insert(final SinkRecord record) {
         metricsJmxReporter.start();
         this.hasInitialized = true;
       }
-      // get offsettoken
-
+      // only get offset token once when service context is initialized
       // ignore ingested files
-      if (record.kafkaOffset() > processedOffset.get()) {
+      // if ingestionDeliveryGuarantee is AT_LEAST_ONCE, ignore offsetPersistedInSnowflake
+      // else discard the record if the record offset is smaller or equal to server side offset
+      if ((ingestionDeliveryGuarantee
+                  == SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.AT_LEAST_ONCE
+              || record.kafkaOffset() > this.offsetPersistedInSnowflake.get())
+          && record.kafkaOffset() > processedOffset.get()) {
```

Comment:
in case of exactly_once, can we set `processedOffset` = `offsetPersistedInSnowflake`?
Then we can avoid this complicated if logic!

Reply:
proceedOffset can sometime be more update to date than the offsetPersistedInSnowflake, so this might not actually work

Reply:
Thank you, the idea here is to make the new code as little as possible to invoke when default/exactly once is set.